angr/state_plugins/symbolizer.py

```diff
@@ -0,0 +1,252 @@
+
+        for _,mo in memory_objects:
+            if not mo.is_bytes and mo.object.symbolic:
+                l.debug("Skipping symbolic memory object %s.", mo)
+                continue
+
+            aligned_base = mo.base - mo.base % (-self.state.arch.bytes)
+            remaining_len = mo.last_addr + 1 - aligned_base
+            if remaining_len < self.state.arch.bytes:
+                continue
+
+            data = mo.bytes_at(aligned_base, remaining_len, allow_concrete=True)
+            if not mo.is_bytes:
+                data = self.state.solver.eval_one(data, cast_to=bytes)
+            #assert self.state.solver.eval_one(storage.load(aligned_base, remaining_len, endness='Iend_BE'), cast_to=bytes).rjust(self.state.arch.bytes) == data
+
+            replacement_content = self._resymbolize_data(
+                data,
+                base=aligned_base,
+                prefix=mo.bytes_at(mo.base, mo.length - remaining_len) if aligned_base != mo.base else b"",
+                skip=() if storage is self.state.memory else (self.state.arch.ip_offset)
+            )
+            if replacement_content is not None:
+                storage.mem.replace_memory_object(mo, replacement_content)
+
+    def resymbolize(self):
+        """
+        Re-symbolizes all pointers in memory. This can be called to symbolize any pointers to target regions
+        that were written (and not mangled beyond recognition) before symbolization was set.
+        """
+        #for i, p_id in enumerate(self.state.registers.mem._pages):
+        #   if i % 100 == 0:
+        #       l.info("%s/%s register pages symbolized", i, len(self.state.registers.mem._pages))
+        #   addr_start = self.state.registers.mem._page_addr(p_id)
+        #   length = self.state.registers.mem._page_size
+        #   self._resymbolize_region(self.state.registers, addr_start, length)
+        #self._resymbolize_region(self.state.registers, self.state.arch.sp_offset, 8)
+
+        for i, p_id in enumerate(self.state.memory.mem._pages):
+            if i % 100 == 0:
+                l.info("%s/%s memory pages symbolized", i, len(self.state.memory.mem._pages))
+            addr_start = self.state.memory.mem._page_addr(p_id)
+            length = self.state.memory.mem._page_size
+            self._resymbolize_region(self.state.memory, addr_start, length)
+
+    @SimStatePlugin.memo
+    def copy(self, memo): # pylint: disable=unused-argument
+        sc = SimSymbolizer()
+        sc._symbolize_all = self._symbolize_all
+        sc.symbolization_target_pages = set(self.symbolization_target_pages)
+        sc.ignore_target_pages = set(self.ignore_target_pages)
+        sc.symbolized_count = self.symbolized_count
+        sc._LE_FMT = self._LE_FMT
+        sc._BE_FMT = self._BE_FMT
+        sc._min_addr = self._min_addr
+        sc._max_addr = self._max_addr
+        sc.page_symbols = dict(sc.page_symbols)
+        return sc
+
+from angr.sim_state import SimState
```

Comment:
Curious: Why not using indirect import?

Reply:
I copied the base code from SimLogger or something. I also like the indirect import better. One sec